index.mjs

```diff
@@ -0,0 +1,437 @@
+  return function(v) {
+    return f (g (v));
+  };
+}
+
+// constant :: a -> b -> a
+export function constant(v) {
+  return function() {
+    return v;
+  };
+}
+
+export function State(run) {
+  if (!(this instanceof State)) return new State (run);
+  this.run = run;
+}
+
+//. ## State
+//.
+//# fantasy-land/of :: a -> State s a
+//.
+//. Replace the value with the given one.
+//.
+//. ```js
+//.   Z.of (State, 1).run (null); // {state: null, value: 1}
+//. ```
+
+State['fantasy-land/of'] = function of(value) {
+  return new State (function(state) {
+    return {state: state, value: value};
+  });
+};
+
+//# modify :: (s -> s) -> State s Null
+//.
+//. Map an old state to a new one. The old state is thrown away.
+//.
+//. ```js
+//.  put (2).run (1); // {state: 2, value: null}
+//. ```
+
+export function modify(f) {
+  return new State (function(state) {
+    return {state: f (state), value: null};
+  });
+}
+
+//# put :: s -> State s Null
+//.
+//. Replace the state with the given one.
+//.
+//. ```js
+//.  put (2).run (1); // {state: 2, value: null}
+//. ```
+
+export function put(state) {
+  return modify (constant (state));
+}
+
+//# get :: State s s
+//.
+//. Replace the returned value of the computation with the state.
+//.
+//. ```js
+//.  Z.chain (() => get, put (1)).run (); // {state: 1, value: 1}
+//. ```
+
+export var get = new State (function(state) {
+  return {state: state, value: state};
+});
+
+//# evalState :: s -> State s a -> a
+//.
+//. Evaluate a state computation with the given initial state and return
+//. the final value, discarding the final state.
+//.
+//. ```js
+//.  evalState () (Z.of (State, 1)); // 1
+//. ```
+
+export function evalState(state) {
+  return function(m) {
+    return m.run (state).value;
+  };
+}
+
+//# execState :: s -> State s a -> s
+//.
+//. Evaluate a state computation with the given initial state and return
+//. the final state, discarding the final value.
+//.
+//. ```js
+//.  execState () (put (1)); // 1
+//. ```
+
+export function execState(state) {
+  return function(m) {
+    return m.run (state).state;
+  };
+}
+
+function next(v) { return {done: false, value: v}; }
+function done(v) { return {done: true, value: v}; }
+
+//# fantasy-land/chainRec :: ((a -> c, b -> c, v) -> State s c, v) -> State s b
+//.
+//. ```js
+//.   Z.chainRec (
+//.     State,
+//.     (next, done, v) => Z.of (State, v > 10 ? done (v) : next (v + 1)),
+//.     1
+//.   ).run (null); // {state: null, value: 11}
+//. ```
+
+State['fantasy-land/chainRec'] = function chainRec(f, v) {
+  return new State (function(state) {
+    var r = {state: state, value: next (v)};
+    while (!r.value.done) {
+      r = f (next, done, r.value.value).run (r.state);
+    }
+    return {state: r.state, value: r.value.value};
+  });
+};
+
+
+//# fantasy-land/chain :: State s a ~> (a -> State s b) -> State s b
+//.
+//. Replace the State with a new one based on the result value of
+//. the previous computation.
```

Comment:
We could just use "Fantasy Land compliant implementation of Chain" (`chain` always does the same, no need to waste words)